File: data-structures/linear/linked-list/singly-linked-list/SinglyLinkedList.java
class Node {
  int data;
  Node next;

  Node(int data) {
    this.data = data;
    this.next = null;
  }
}

public class SinglyLinkedList {
  private Node head;
  private int size;

  public SinglyLinkedList() {
    head = null;
    size = 0;
  }

  // -------------Core-Operations-------------
  public boolean isEmpty() {
    return (size == 0);
  }

  public int size() {
    return size;
  }

  public int getAtIndex(int index) {
    int index_counter = 0;
    Node temp = head;
    while (temp != null) {
      if (index == index_counter)
        return temp.data;
      temp = temp.next;
      index_counter++;
    }
    return -1;
  }

  public boolean search(int value) {
    Node curr = head;

    while (curr != null) {
      if (curr.data == value) {
        return true;
      }

      curr = curr.next;
    }

    return false;
  }

  public void Traverse() {
    Node curr = head;
    while (curr != null) {
      System.out.print(curr.data);
      if (curr.next != null)
        System.out.print(" -> ");
      curr = curr.next;
    }
    System.out.println("NULL");
  }

  // -------------Insertions-------------

  public void insertFront(int data) {
    Node newNode = new Node(data);

    newNode.next = head;
    head = newNode;
    size++;
  }

  public void insertLast(int data) {
    Node newNode = new Node(data);

    if (head == null) {
      head = newNode;
    } else {
      Node curr = head;

      while (curr.next != null) {
        curr = curr.next;
      }

      curr.next = newNode;
    }

    size++;
  }

  public void insertAtIndex(int index, int data) {
    if (index < 0 || index > size) {
      System.out.println("Index should be between 0 and " + (size - 1));
      return;
    }

    if (index == 0) {
      insertFront(data);
      return;
    }

    Node newNode = new Node(data);
    Node currNode = head;

    for (int i = 0; i < index - 1; i++) {
      currNode = currNode.next;
    }

    newNode.next = currNode.next;
    currNode.next = newNode;
    size++;
  }

  // -------------Deletions-------------

  public void deleteFirst() {
    if (head == null) {
      System.out.println("List is empty");
      return;
    }

    head = head.next;
    size--;
  }

  public void deleteLast() {
    if (head == null) {
      System.out.println("List is empty");
      return;
    }

    if (head.next == null) {
      head = null;
    } else {
      Node curr = head;

      while (curr.next.next != null) {
        curr = curr.next;
      }

      curr.next = null;
    }

    size--;
  }

  public void deleteAt(int index) {
    if (index < 0 || index >= size) {
      System.out.println("Index should be between 0 and " + (size - 1));
      return;
    }

    if (index == 0) {
      deleteFirst();
      return;
    }

    Node currNode = head;

    for (int i = 0; i < index - 1; i++) {
      currNode = currNode.next;
    }

    currNode.next = currNode.next.next;
    size--;
  }

  public void deleteKey(int value) {
    if (head == null) {
      System.out.println("List is empty");
      return;
    }

    if (head.data == value) {
      head = head.next;
      size--;
      return;
    }

    Node currNode = head;

    while (currNode.next != null && currNode.next.data != value) {
      currNode = currNode.next;
    }

    if (currNode.next == null) {
      System.out.println("Value " + value + " is not found in the list");
      return;
    }

    currNode.next = currNode.next.next;
    size--;
  }
  // -------------Utilities-------------

  public void Update(int update_val, int new_data) {
    Node temp = head;
    while (temp != null && temp.data != update_val) {
      temp = temp.next;
    }
    if (temp == null) {
      System.out.println("Node with given data: " + update_val + " not found");
      return;
    }
    temp.data = new_data;
    System.out.println("Data updated from " + update_val + " to " + new_data);
  }

  public void findMiddle() {
    if (head == null) {
      System.out.println("List is empty");
      return;
    }
    Node fast = head;
    Node slow = head;
    while (fast != null && fast.next != null) {
      slow = slow.next;
      fast = fast.next.next;
    }
    System.out.println("The middle value is: " + slow.data);
  }

  public void Reverse() {
    if (head == null)
      return;
    Node prev_Node = null;
    Node current_Node = head;
    Node next_Node = null;

    while (current_Node != null) {
      next_Node = current_Node.next;
      current_Node.next = prev_Node;
      prev_Node = current_Node;
      current_Node = next_Node;
    }
    head = prev_Node;
  }

  public static void main(String[] args) {
    SinglyLinkedList list = new SinglyLinkedList();

    list.insertLast(10);
    list.insertLast(20);
    list.insertLast(30);
    list.Traverse(); // 10 -> 20 -> 30 -> NULL

    list.insertFront(5);
    list.Traverse(); // 5 -> 10 -> 20 -> 30 -> NULL

    list.insertAtIndex(2, 15);
    list.Traverse(); // 5 -> 10 -> 15 -> 20 -> 30 -> NULL

    list.insertAtIndex(3, 18);
    list.Traverse(); // 5 -> 10 -> 15 -> 18 -> 20 -> 30 -> NULL

    System.out.println("Size: " + list.size());
    System.out.println("Get(2): " + list.getAtIndex(2));

    list.Update(15, 16);
    list.Traverse(); // 5 -> 10 -> 16 -> 18 -> 20 -> 30 -> NULL

    list.deleteFirst();
    list.Traverse(); // 10 -> 16 -> 18 -> 20 -> 30 -> NULL

    list.deleteLast();
    list.Traverse(); // 10 -> 16 -> 18 -> 20 -> NULL

    list.deleteAt(1);
    list.Traverse(); // 10 -> 18 -> 20 -> NULL

    list.deleteKey(18);
    list.Traverse(); // 10 -> 20 -> NULL

    list.Reverse();
    list.Traverse(); // 20 -> 10 -> NULL

    list.insertLast(25);
    list.insertLast(35);
    list.Traverse(); // 20 -> 10 -> 25 -> 35 -> NULL

    list.findMiddle(); // prints "The middle value is: 10" or "25"
  }

}

File: data-structures/linear/linked-list/singly-linked-list/SinglyLinkedList.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
  int data;
  struct Node *next;
};

struct Node *createNode(int value) {
  struct Node *newNode = (struct Node *)malloc(sizeof(struct Node));
  newNode->data = value;
  newNode->next = NULL;
  return newNode;
}

int getSize(struct Node *head) {
  int count = 0;
  while (head != NULL) {
    count++;
    head = head->next;
  }
  return count;
}

void traverse(struct Node *head) {
  while (head != NULL) {
    printf("%d -> ", head->data);
    head = head->next;
  }
  printf("NULL\n");
}

struct Node *update(struct Node *head, int updateValue, int newData) {
  struct Node *tempNode = head;
  while (tempNode != NULL && tempNode->data != updateValue)
    tempNode = tempNode->next;

  if (tempNode == NULL) {
    printf("Node with given data: %d not found\n", updateValue);
    return head;
  }
  tempNode->data = newData;
  return head;
}

void search(struct Node *head, int searchData) {
  struct Node *tempNode = head;
  int pos = 0;
  while (tempNode != NULL) {
    if (tempNode->data == searchData) {
      printf("Data %d found at position %d\n", tempNode->data, pos);
      return;
    }
    tempNode = tempNode->next;
    pos++;
  }
  printf("Data %d not found\n", searchData);
}

int getAtIndex(struct Node *head, int index) {
  int indexCounter = 0;
  while (head != NULL) {
    if (index == indexCounter)
      return head->data;
    head = head->next;
    indexCounter++;
  }
  printf("Index out of range\n");
  return -1;
}

void isEmpty(struct Node *head) {
  printf("List %s empty\n", head ? "isn't" : "is");
}

struct Node *insertFirst(struct Node *head, int value) {
  struct Node *newNode = createNode(value);
  newNode->next = head;
  head = newNode;
  return head;
}

struct Node *insertLast(struct Node *head, int value) {
  struct Node *newNode = createNode(value);
  if (head == NULL)
    return newNode;
  struct Node *tempNode = head;
  while (tempNode->next != NULL)
    tempNode = tempNode->next;
  tempNode->next = newNode;
  return head;
}

struct Node *insertAt(struct Node *head, int position, int value) {
  if (position == 0)
    return insertFirst(head, value);

  struct Node *tempNode = head;
  for (int i = 0; i < position - 1 && tempNode != NULL; i++)
    tempNode = tempNode->next;

  if (tempNode == NULL) {
    printf("Position out of range\n");
    return head;
  }

  struct Node *newNode = createNode(value);
  newNode->next = tempNode->next;
  tempNode->next = newNode;
  return head;
}

struct Node *insertAfter(struct Node *head, int NodeData, int data) {
  struct Node *newNode = createNode(data);
  if (head == NULL) {
    return newNode;
  }
  struct Node *tempNode = head;
  while (tempNode != NULL && tempNode->data != NodeData) {
    tempNode = tempNode->next;
  }
  if (tempNode == NULL) {
    printf("Node with given data: %d not found\n", NodeData);
    return head;
  }
  newNode->next = tempNode->next;
  tempNode->next = newNode;
  return head;
}

struct Node *deleteFirst(struct Node *head) {
  if (head == NULL)
    return NULL;

  struct Node *tempNode = head;
  head = head->next;
  printf("Deleted value: %d\n", tempNode->data);
  free(tempNode);
  return head;
}

struct Node *deleteLast(struct Node *head) {
  if (head == NULL)
    return NULL;
  if (head->next == NULL) {
    printf("Deleted value: %d\n", head->data);
    free(head);
    return NULL;
  }
  struct Node *tempNode = head;
  while (tempNode->next->next != NULL)
    tempNode = tempNode->next;
  printf("Deleted value: %d\n", tempNode->next->data);
  free(tempNode->next);
  tempNode->next = NULL;
  return head;
}

struct Node *deleteAt(struct Node *head, int position) {
  if (head == NULL)
    return NULL;
  if (position == 0)
    return deleteFirst(head);

  struct Node *tempNode = head;
  for (int i = 0; i < position - 1 && tempNode->next != NULL; i++)
    tempNode = tempNode->next;

  if (tempNode->next == NULL) {
    printf("Position out of range\n");
    return head;
  }

  struct Node *toDelete = tempNode->next;
  tempNode->next = tempNode->next->next;
  printf("Deleted value: %d\n", toDelete->data);
  free(toDelete);
  return head;
}

struct Node *deleteKey(struct Node *head, int deleteData) {
  if (head == NULL)
    return NULL;
  if (head->data == deleteData) {
    struct Node *tempNode = head;
    head = head->next;
    printf("Deleted value: %d\n", tempNode->data);
    free(tempNode);
    return head;
  }

  struct Node *prev = head;
  struct Node *toDelete = head->next;

  while (toDelete != NULL && toDelete->data != deleteData) {
    prev = toDelete;
    toDelete = toDelete->next;
  }
  if (toDelete == NULL) {
    printf("Node with given data: %d not found\n", deleteData);
    return head;
  }
  prev->next = toDelete->next;
  printf("Deleted value: %d\n", toDelete->data);
  free(toDelete);
  return head;
}

struct Node *reverse(struct Node *head) {
  if (head == NULL)
    return NULL;
  struct Node *prevNode = NULL;
  struct Node *currentNode = head;
  struct Node *nextNode = NULL;

  while (currentNode != NULL) {
    nextNode = currentNode->next;
    currentNode->next = prevNode;
    prevNode = currentNode;
    currentNode = nextNode;
  }
  head = prevNode;
  return head;
}

int last(struct Node *head) {
  if (head == NULL) {
    printf("List empty\n");
    return -1;
  }
  while (head->next != NULL)
    head = head->next;
  return head->data;
}

struct Node *copy(struct Node *head, struct Node *headCopy) {
  if (headCopy != NULL) {
    printf("Warning: destination list not empty. Copy may append to existing "
           "list.\n");
  }
  struct Node *tempNode = head;
  while (tempNode != NULL) {
    headCopy = insertLast(headCopy, tempNode->data);
    tempNode = tempNode->next;
  }
  return headCopy;
}

struct Node *findMiddle(struct Node *head) {
  if (head == NULL) {
    printf("List is empty\n");
    return;
  }
  struct Node *fast = head;
  struct Node *slow = head;
  while (fast != NULL && fast->next != NULL) {
    slow = slow->next;
    fast = fast->next->next;
  }
  return slow;
}
